lib/compress/zstdmt_compress.c

```diff
@@ -664,13 +666,29 @@ static ZSTD_CCtx_params ZSTDMT_initJobCCtxParams(ZSTD_CCtx_params const params)
     return jobParams;
 }
 
-/* ZSTDMT_getNbThreads():
+/*! ZSTDMT_updateCParams_whileCompressing() :
+ *  Update compression level and parameters (except wlog)
+ *  while compression is ongoing.
+ *  New parameters will be applied to next compression job. */
+void ZSTDMT_updateCParams_whileCompressing(ZSTDMT_CCtx* mtctx, int compressionLevel, ZSTD_compressionParameters cParams)
+{
+    U32 const saved_wlog = mtctx->params.cParams.windowLog;   /* Do not modify windowLog while compressing */
+    DEBUGLOG(5, "ZSTDMT_updateCParams_whileCompressing (level:%i)",
+                compressionLevel);
+    mtctx->params.compressionLevel = compressionLevel;
+    if (compressionLevel != ZSTD_CLEVEL_CUSTOM)
+        cParams = ZSTD_getCParams(compressionLevel, mtctx->frameContentSize, 0 /* should be dictSize */ );
```

Comment:
What do you mean `should be dictSize`? Are you labeling the parameter, or is it a TODO?

Reply:
It's a reminder that the strange corner case scenario of multi-threaded compression started from `ZSTD_compress_generic()` using a `CDict` is not very clear to me regarding compression parameters.

When there is a dictionary, it's only active for first job.
Job 1 compression parameters are enforced from `CDict`.
But since Job 1 writes the frame header, it's necessary to preserve compression parameters across all jobs, so how are they transferred ?

When invoking ZSTDMT interface directly, situation is clearer : compression parameters are extracted from `CDict`, and then passed as parameter to `ZSTD_initCStream_internal()` : 
https://github.com/facebook/zstd/blob/dev/lib/compress/zstdmt_compress.c#L975

When invoked from `ZSTD_compress_generic()`, this is less clear to me : 
initialisation invokes `ZSTD_initCStream_internal()`, but `params` seem to come from `requestedParams`, and I'm not sure if this is correctly synchronised with `CDict`.
https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L3177

So still some clarifications to work out.

Reply:
OK, the way it's setup currently : 
when there is a `CDict` set, either through `ZSTD_CCtx_refCDict()` or `ZSTD_CCtx_loadDictionary()`, it's no longer possible to change any compression parameter.

So, currently, changing compression parameters "on the fly" is only possible for compression without dictionary.

That's a limitation, but not a strong one : dictionary and multi-threading have no reason to be triggered together. Data flow which can benefit from automatic compression level adaptation are large, so they don't benefit from dictionary.